crates/ty_python_semantic/src/types/infer/builder.rs

```diff
@@ -1945,28 +1947,34 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
                 return;
             }
 
-            for invalid in self
-                .return_types_and_ranges
+            for (invalid_ty, range) in self
+                .returnees
                 .iter()
                 .copied()
-                .filter_map(|ty_range| match ty_range.ty {
-                    // We skip `is_assignable_to` checks for `NotImplemented`,
-                    // so we remove it beforehand.
-                    Type::Union(union) => Some(TypeAndRange {
-                        ty: union.filter(self.db(), |ty| !ty.is_notimplemented(self.db())),
-                        range: ty_range.range,
-                    }),
-                    ty if ty.is_notimplemented(self.db()) => None,
-                    _ => Some(ty_range),
+                .filter_map(|returnee| {
+                    match returnee
+                        .expression
+                        .map_or(Type::none(self.db()), |expression| {
```

Comment:
Why would we fallback to `Type::none` here?

It seems like we should just `expect` here, we should always have inferred a type for every returnee expression.